src/common/sv2_noise.h

```diff
@@ -0,0 +1,298 @@
+ */
+class Sv2CipherState
+{
+public:
+    Sv2CipherState() = default;
+    explicit Sv2CipherState(NoiseHash&& key);
+
+    /** Decrypt message
+     * @param[in] associated_data associated data
+     * @param[in] ciphertext message with encrypted and authenticated chunks.
+     * @param[out] plain message (defragmented)
+     * @returns whether decryption succeeded
+     */
+    [[nodiscard]] bool DecryptWithAd(Span<const std::byte> associated_data, Span<std::byte> ciphertext, Span<std::byte> plain);
+
+    /** Encrypt message
+     * @param[in] associated_data associated data
+     * @param[in] plain message
+     * @param[out] ciphertext message with encrypted and authenticated chunks.
+     * @returns whether decryption succeeded
+     */
+    [[nodiscard]] bool EncryptWithAd(Span<const std::byte> associated_data, Span<const std::byte> plain, Span<std::byte> ciphertext);
+
+    /** The message will be chunked in NOISE_MAX_CHUNK_SIZE parts and expanded
+     *  by 16 bytes per chunk for its MAC.
+     *
+     * @param[in] plain message. Can't point to the same memory location as ciphertext,
+     *                  because each encrypted message chunk would override the
+     *                  start of the next plain text chunk.
+     * @param[out] ciphertext   message with encrypted and authenticated chunks
+     * @return whether encryption succeeded. Only fails if nonce is uint64_max.
+     */
+    [[nodiscard]] bool EncryptMessage(Span<const std::byte> plain, Span<std::byte> ciphertext);
+
+    /** Decrypt message.
+     *
+     * @param[in] ciphertext encrypted message
+     * @param[out] plain decrypted message. May point to the same memory location
+     *                  as ciphertext. The result is defragmented.
+     */
+    [[nodiscard]] bool DecryptMessage(Span<std::byte> ciphertext, Span<std::byte> plain);
+
+private:
+    NoiseHash m_key{0};
+    uint64_t m_nonce = 0;
+};
+
+/*
+ * A SymmetricState object contains a CipherState plus m_chaining_key (ck) and
+ * m_hash_output (h) variables. It is so-named because it encapsulates all the
+ * "symmetric crypto" used by Noise. During the handshake phase each party has
+ * a single SymmetricState, which can be deleted once the handshake is finished.
+ *
+ * See chapter "5. Processing rules" of the Noise paper.
+ */
+class Sv2SymmetricState
+{
+public:
+    // Sha256 hash of the ascii encoding - "Noise_NX_Secp256k1+EllSwift_ChaChaPoly_SHA256".
+    // This is the first step required when setting up the chaining key.
+    static constexpr NoiseHash PROTOCOL_NAME_HASH = {
+        46, 180, 120, 129, 32, 142, 158, 238, 31, 102, 159, 103, 198, 110, 231, 14,
+        169, 234, 136, 9, 13, 80, 63, 232, 48, 220, 75, 200, 62, 41, 191, 16};
+
+    // The double hash of protocol name "Noise_NX_EllSwiftXonly_ChaChaPoly_SHA256".
+    static constexpr NoiseHash PROTOCOL_NAME_DOUBLE_HASH = {146, 47, 163, 46, 79, 72, 124, 13, 89, 202, 163, 190, 215, 137, 156, 227,
```

Comment:
If you consider updating for other reasons, could you format this in the same way as PROTOCOL_NAME_HASH?